Hchow/tlt 2633 class roster selenium

hktest opened the pull request:
Setting up selenium tests for the new class roster tool. 
Things to note in this PR: 
1.  There is a skeleton permission_tests in place  class_roster_permission_tests.  This is just a skeleton for when we get specifications from the product owners into who should have access to this tool, at which point, the test_data spreadsheet could just be modified without modifying the underlying code.  
2.  Tested and ran regression_suite.py for the canvas_course_account_admin successfully.

hktest committed:
Building out the selenium page object structure for the new class roster tool.

hktest committed:
- cleaned up code, fix up pep8 issues, added permission tests placeholder (actual permission not yet provided), added test data.

hktest committed:
Adding in additional test logic.

hktest committed:
refactored and fixed TODOs.

hktest committed:
Added test to test that class roster URL link contains a class number that matches the test course

elliottyates commented on canvas_course_admin_tools/settings/local.py:
```diff
@@ -29,6 +29,12 @@
 dictConfig(LOGGING)
 
 SELENIUM_CONFIG = {
+    'class_roster': {
```
maybe after canvas_base_url for alphabetical consistency?

hktest replied:
Fixed!

elliottyates commented on canvas_course_admin_tools/settings/local.py:
```diff
@@ -29,6 +29,12 @@
 
 SELENIUM_CONFIG = {
+    'class_roster': {
+        'course_link': 'https://canvas.dev.tlt.harvard.edu/courses/3787'
```
maybe this should be built from the canvas_base_url in the code, like the other links (see manage_course's `relative_url`, below)? That way any change to the base URL in the future (including running selenium tests on qa, stage, etc) will be picked up automatically.

hktest replied:
Good point. I went this way because I thought it would be easier than do URL join in the code, but you're right - it makes for cleaner and more consistent code.  Fixing this -- thanks!

elliottyates commented on selenium_tests/class_roster/class_roster_base_test_case.py:
```diff
@@ -0,0 +1,49 @@
+        # instantiate
+        self.dashboard_page = CourseAdminDashboardPage(self.driver)
+        self.class_roster_page = MainPageObject(self.driver)
+        self.main_page = MainPageObject(self.driver)
```
duplicate PO -- we already have it from line 27. We can collapse these two (and fix the references e.g. on line 38 and 49).

hktest replied:
Fixed!

elliottyates commented on selenium_tests/class_roster/class_roster_base_test_case.py:
```diff
@@ -0,0 +1,49 @@
+        self.assertTrue(self.class_roster_page.is_loaded())
+
+
+        # initialize
```
This looks like duplicate code -- all this code below here is already done by line 38, right?

hktest replied:
Discussed.  Cleaning out the following unnecessary code.  Thanks!

elliottyates commented on selenium_tests/class_roster/class_roster_permission_tests.py:
```diff
@@ -0,0 +1,43 @@
+                    .class_roster_button_is_displayed(),
+                'User {} should have access to see tool but does not'.format(
+                    user_id)
+            )
```
Do we also want to check that the user can actually get into the tool?

hktest replied:
That seems redundant, no?   If a user does not have access to the tool, they wouldn't be able to see the class roster button on page (this is checked by the "class_roster_button_is_displayed()" method).  So if they can't see the button, they won't have any option to get into the tool.

elliottyates replied:
@hktest If they _do_ see the button but when they click on it they are denied (they see an error), that would be a bug. We may want to check for that, since it's just another line or two of code.

hktest replied:
@elliottyates: ah, got it.  I see what you mean now.  I will add in that check, thanks!  :+1:

elliottyates commented on selenium_tests/class_roster/class_roster_permission_tests.py:
```diff
@@ -0,0 +1,43 @@
+
+        else:
+            raise ValueError(
+                'given_access column for user {} must be either \'yes\' or '
```
fyi you can avoid all the backslashes by wrapping the string in double-quotes. (http://learnpythonthehardway.org/book/ex10.html)

hktest replied:
Fixed, and thanks for the reference.

elliottyates commented on selenium_tests/class_roster/class_roster_tests.py:
```diff
@@ -0,0 +1,25 @@
+        link_url = self.main_page.get_link_url(course_name_display)
+
+        # Verify that the expect course number appears in the link URL
+        self.assertTrue(url_link_course_number in link_url,
```
Do we also want to check the other dynamic parameters (user ID and term) are correct?

hktest replied:
Great idea.  I thought I start off with course_number first.  I'll add in a check for the term as well, since that is tied in with the course.  For the robustness of the test, I'll skip the user_id check if/when the test user gets changed.

hktest replied:
Fixed.  Added another assertion here.

elliottyates commented on selenium_tests/class_roster/page_objects/class_roster_main_page.py:
```diff
@@ -0,0 +1,34 @@
+
+
+class Locators(object):
+    CLASS_ROSTER_BREADCRUMB = (By.XPATH, './/h1[contains(., "Class Roster")]')
```
This will be true for any child pages, because it will always be present in the breadcrumbs. Currently the app has no child pages, so this is fine, but we might want to fix it to avoid subtle bugs in the future if the tool changes.

hktest replied:
@elliottyates, hmmm, I seemed to be stuck on how to fix this. Unlike the other tests which have a exact match on "manage course", this is part of the breadcrumbs it's not an exact match because the full breadcrumb nav also includes "manage course > "  Tried against h1 and nav tags and tried using the console, but I haven't been successful yet.   I'll have to dig a bit more on this one, but may have to pick your brains on it next week.

elliottyates commented on selenium_tests/class_roster/page_objects/class_roster_main_page.py:
```diff
@@ -0,0 +1,34 @@
+
+    def get_link_url(self, text_value):
+        """
+        Returns the URL from the partial link text passed in from tests.
```
Maybe a little more descriptive? E.g. about what the links are/represent/do in the context of the page and workflow, and why you'd want to get the URL instead of clicking on them.

hktest replied:
Ah, good idea.  Fixed.  Adding a more detailed description.

elliottyates commented on selenium_tests/course_admin/page_objects/course_admin_dashboard_page_object.py:
```diff
@@ -6,6 +6,7 @@
 
 
 class Locators(object):
+    CLASS_ROSTER_BUTTON = (By.ID, "course-roster")
```
(Darn it! :) We forgot to change the ID from `course-roster` to `class-roster`!)

hktest replied:
easy change, if and when it gets changed. not a big deal here.  ;)

elliottyates commented on selenium_tests/course_admin/page_objects/course_admin_dashboard_page_object.py:
```diff
@@ -26,9 +27,30 @@ def manage_people_button_is_displayed(self):
+        """
+        try:
+            self.focus_on_tool_frame()
+            self.find_element(*Locators.CLASS_ROSTER_BUTTON)
```
This only checks the DOM for the button; it might return the element even if the element is hidden. I think we want something like http://selenium-python.readthedocs.io/api.html#selenium.webdriver.remote.webelement.WebElement.is_displayed after getting the element reference as well.

hktest replied:
Great catch - can't believe I missed this.  Thanks and fixed!

elliottyates commented:
Some duplicate code, and some suggestions for more robust tests. Otherwise 👍

hktest committed:
Made changes based on PR feedback.
A few more changes still to be completed.

hktest committed:
- Added additional logic to the permission test per PR feedback.
- Fixed up some incorrect changes - added "self" back in to base_test_case so I can call them from the test.

hktest committed:
Fixed per PR feedback.  Added in a check for term id in the URL.

hktest committed:
Fixed - the assertion needs to check against term code

hktest committed:
adding in note to describe locator limitation

hktest committed:
Merge branch 'develop' into hchow/tlt_2633_class_roster_selenium
